### PR TITLE
Add language statistics API endpoint

### LINKED ISSUE #10232: [Feature] [API] export repo-language info
since #8037 created `LanguageStat`'s for each repo ... get this info via API too

### PR DESCRIPTION
**Pull request by lafriks**
Fixes #10232

### CODE REVIEW

File: routers/api/v1/repo/language.go
Comment (6543) [inline routers/api/v1/repo/language.go:19]: pleace move this convertion into modules/convert

**Review by 6543 — CHANGES_REQUESTED** [review]

File: routers/api/v1/repo/language.go
Comment (lafriks) [reply routers/api/v1/repo/language.go:19]: It's not really conversation

Comment (6543) [reply routers/api/v1/repo/language.go:19]: @lunny what do you think?
at the moment I think it is a convert and lafriks dont

so the 3rd will decide ;)

Comment (lafriks) [reply routers/api/v1/repo/language.go:19]: In convert module there are just convertation between types. In this case it is specific internal type serialziation/encoding to json. That would mean also moving that languageResponse type definition to convert module that is not really place for it.

**Comment by CirnoT** [comment]
Why use custom and not `json`?

**Comment by 6543** [comment]
I dont know how hard this is - but tests would be nice

**Comment by lafriks** [comment]
@CirnoT to return languages in descending order by respective code size in bytes

**Comment by CirnoT** [comment]
JSON map objects have undefined order, no one should depend on what order we return them in - if they do, their code is bound to break on something at some point, 100% guranteed.

**Comment by lafriks** [comment]
that's why I don't use map and use custom json serialization

**Comment by CirnoT** [comment]
You don't understand me, it's not Go spec, its JSON spec. You shall not return ordered or depend on order of JSON objects, EVER.

In fact you can not even rely on what order you will get after you use some language built-in function to fetch that JSON string from our API endpoint.

**Comment by CirnoT** [comment]
In other words, If you have code like this:

```js
JSON.parse('{"PHP":23334, "Go":6754}')
```
And rely or expect that PHP will always be first on iteration, you are doing things wrong.

**Comment by CirnoT** [comment]
The reason why GitHub returns them ordered is most likely because they internally iterate in that way, not because they TRY to present such order.

**Comment by lafriks** [comment]
But if you use JSON reader like parsing than you will get primary language first. imho github specifically returns them it this order not just by coincidence

**Comment by lafriks** [comment]
Test added

**Comment by CirnoT** [comment]
I can not agree with explicitly suggesting order and encouraging developers to rely on it; it's just wrong.

**Comment by CirnoT** [comment]
>But if you use JSON reader like parsing than you will get primary language first

Today it works for you, tomorrow you will update unrelated dependency and it won't work anymore because you were relying on something that is strictly against JSON spec.

**Comment by lafriks** [comment]
It's not against spec, against spec would be if json was invalid. There is no reason why this code would start returning them in different order. There are languages that has ordered map classes that would decode this structure just fine, if you just use default JSON.parse it will work same no matter in what order we return them

**Comment by CirnoT** [comment]
https://www.json.org/json-en.html
>An object is an unordered set of name/value pairs

**Comment by CirnoT** [comment]
Your code attempts to explicitly return object in specific order, and by your explanation I fully expect that you believe you can write code that will assume first element to be always primary language. That is incorrect. Some libraries will preserve input order, some will re-arrange it alphabetically, others will try to move integer values above other values.

I don't mind any specific order, what I have issue is that you are forcing specific order and to do that, you go to insane lengths of writing own `MarshalJSON`.

**Comment by CirnoT** [comment]
For example, if someone were to naively use our API in Go and expect first element to by primary language, he would be very surprised by the result: https://play.golang.org/p/XYUrtJlEG_b

As such, I don't see any point on keeping any order for what we return, especially when it requires additional code that is subject to breakage.

**Comment by lafriks** [comment]
I understand all that but that does not mean we can not return it using same order as github to be as compatible with them as possible. It's up to the API consumer to either to depend on it or not

**Comment by lafriks** [comment]
As for MarshalJSON length it's not that insanely big or complex

**Comment by CirnoT** [comment]
>to be as compatible with them as possible

It's compatible either way

>It's up to the API consumer to either to depend on it or not

No, it's not. Attempting to depend on it is incorrect.

**Comment by lafriks** [comment]
I do understand your point but I don't see much of the problem returning exactly like github does and I can bet they do that order intentionally

**Review by 6543 — APPROVED** [review]
beside the none default way how it is converted ... I'm ok with this

works smothly 🙃

**Comment by CirnoT** [comment]
I am not convinced that your arguments warrant introducing custom JSON marshal function; I'll let someone else review it.

**Review by zeripath — APPROVED** [review]
Although I take @CirnoT's point that if we want order we should use a list - I suspect that we need to do this to keep compatibility with GH.

**Comment by techknowlogick** [comment]
Like @CirnoT I also would prefer the built-in marshal, and that even though Github's list is sorted by us doing it then that creates a guarantee that the first elem will always be primary, however Github doesn't state that as a promise in their documentation. So I don't think we should make that promise either to encourage devs working with API to not take data at face value.

If we put it in without the sorting, then later we decide to presort it, then that wouldn't be a breaking change, but if we start out sorting the list and decide to switch that would be breaking.

Otherwise, this PR LGTM and so I won't block it. Sidenote: @lafriks I can click "update branch" are you able to check "accept edits from maintainers"?

**Comment by lafriks** [comment]
I can't seem to find such option for existing PR :)

**Comment by 6543** [comment]
Ping bot

**Pull request merged**
